internalshared/configutil/userlockout.go

```diff
@@ -0,0 +1,175 @@
+	LockoutCounterResetRaw interface{}   `hcl:"lockout_counter_reset"`
+	DisableLockout         bool          `hcl:"-"`
+	DisableLockoutRaw      interface{}   `hcl:"disable_lockout"`
+}
+
+func ParseUserLockouts(result *SharedConfig, list *ast.ObjectList) error {
+	var err error
+	result.UserLockouts = make([]*UserLockout, 0, len(list.Items))
+	userLockoutsMap := make(map[string]*UserLockout)
+	for i, item := range list.Items {
+		var userLockoutConfig UserLockout
+		if err := hcl.DecodeObject(&userLockoutConfig, item.Val); err != nil {
+			return multierror.Prefix(err, fmt.Sprintf("userLockouts.%d:", i))
+		}
+
+		// Base values
+		{
+			switch {
+			case userLockoutConfig.Type != "":
+			case len(item.Keys) == 1:
+				userLockoutConfig.Type = strings.ToLower(item.Keys[0].Token.Value().(string))
+			default:
+				return multierror.Prefix(errors.New("auth type for user lockout must be specified, if it applies to all auth methods specify \"all\" "), fmt.Sprintf("user_lockouts.%d:", i))
+			}
+
+			userLockoutConfig.Type = strings.ToLower(userLockoutConfig.Type)
+			// Supported auth methods for user lockout configuration: ldap, approle, userpass
+			// "all" is used to apply the configuration to all supported auth methods
+			switch userLockoutConfig.Type {
+			case "all", "ldap", "approle", "userpass":
+				result.found(userLockoutConfig.Type, userLockoutConfig.Type)
+			default:
+				return multierror.Prefix(fmt.Errorf("unsupported auth type %q", userLockoutConfig.Type), fmt.Sprintf("user_lockouts.%d:", i))
+			}
+		}
+
+		// Lockout Parameters
+		{
+			if userLockoutConfig.LockoutThresholdRaw != nil {
+				if userLockoutConfig.LockoutThreshold, err = parseutil.ParseInt(userLockoutConfig.LockoutThresholdRaw); err != nil {
+					return multierror.Prefix(fmt.Errorf("error parsing lockout_threshold: %w", err), fmt.Sprintf("user_lockouts.%d", i))
+				}
+			}
+
+			if userLockoutConfig.LockoutDurationRaw != nil {
+				if userLockoutConfig.LockoutDuration, err = parseutil.ParseDurationSecond(userLockoutConfig.LockoutDurationRaw); err != nil {
+					return multierror.Prefix(fmt.Errorf("error parsing lockout_duration: %w", err), fmt.Sprintf("user_lockouts.%d", i))
+				}
+				if userLockoutConfig.LockoutDuration < 0 {
+					return multierror.Prefix(errors.New("lockout_duration cannot be negative"), fmt.Sprintf("user_lockouts.%d", i))
+				}
+
+			}
+
+			if userLockoutConfig.LockoutCounterResetRaw != nil {
+				if userLockoutConfig.LockoutCounterReset, err = parseutil.ParseDurationSecond(userLockoutConfig.LockoutCounterResetRaw); err != nil {
+					return multierror.Prefix(fmt.Errorf("error parsing lockout_counter_reset: %w", err), fmt.Sprintf("user_lockouts.%d", i))
+				}
+				if userLockoutConfig.LockoutCounterReset < 0 {
+					return multierror.Prefix(errors.New("lockout_counter_reset cannot be negative"), fmt.Sprintf("user_lockouts.%d", i))
+				}
+
+			}
+			if userLockoutConfig.DisableLockoutRaw != nil {
```

Comment:
Yes, I agree that the usual way is to set nil for 'raw' values as soon as they are parsed. The issue with checking non-raw entries is because of fields lockout threshold (uint) and disable lockout (bool). It will be difficult to figure out if the user configured lockout threshold with value 0 or if the user did not configure this value at all. In case of other configurations like Listeners, we append values after the parse but in this case, we parse through the map values to set default values for each "type" mentioned in config in result.UserLockouts.

Reply:
One way to get away from this is set these values to defaults before parsing raw values. But, I guess then there are values that are populated by `all` type.